thread_pthread_mn.c

```diff
@@ -750,14 +861,51 @@ timer_thread_polling(rb_vm_t *vm)
             // simply retry
             break;
           default:
-            perror("epoll_wait");
-            rb_bug("epoll_wait errno:%d", errno);
+            perror("kq");
+            rb_bug("kq errno:%d", errno);
         }
         break;
 
       default:
         RUBY_DEBUG_LOG("%d event(s)", r);
 
+#if HAVE_SYS_EVENT_H
+        for (int i=0; i<r; i++) {
+            rb_thread_t *th = (rb_thread_t *)timer_th.finished_events[i].udata;
+            int fd = (int)timer_th.finished_events[i].ident;
+            int16_t filter = timer_th.finished_events[i].filter;
+
+            if (th == NULL) {
+                // wakeup timerthread
+                RUBY_DEBUG_LOG("comm from fd:%d", timer_th.comm_fds[1]);
+                consume_communication_pipe(timer_th.comm_fds[0]);
+            } else {
+                // wakeup specific thread by IO
+                RUBY_DEBUG_LOG("io event. wakeup_th:%u event:%s%s",
+                                rb_th_serial(th),
+                                (filter == EVFILT_READ) ? "read/" : "",
+                                (filter == EVFILT_WRITE) ? "write/" : "");
+
+                rb_native_mutex_lock(&timer_th.waiting_lock);
+                {
+                    if (th->sched.waiting_reason.flags) {
+                        // delete from chain
+                        ccan_list_del_init(&th->sched.waiting_reason.node);
+                        timer_thread_unregister_waiting(th, fd, kqueue_translate_filter_to_flags(filter));
+
+                        th->sched.waiting_reason.flags = thread_sched_waiting_none;
+                        th->sched.waiting_reason.data.fd = -1;
+                        th->sched.waiting_reason.data.result = filter;
```

Comment:
We could receive multiple filters back, so setting this over and over again and ending on the last value _probably_ doesn't make sense.

Reply:
It's actually fine for now. Because the flags get set back to `thread_sched_waiting_none`, any subsequent event in the loop pointing at this thread will act like a no-op and hit the `// already released` `else` clause. It is not granular about which event fired, just that the thread can be active again.